src/trajectory.jl

```diff
@@ -13,6 +13,70 @@ transition(
     z::PhasePoint
 ) where {I<:AbstractIntegrator} = transition(GLOBAL_RNG, τ, h, z)
 
+#################################
+### Hong's abstraction starts ###
+#################################
+
+###
+### Create a `Termination` type for each `Trajectory` type, e.g. HMC, NUTS etc.
+### Merge all `Trajectory` types, and make `transition` dispatch on `Termination`,
+### such that we can overload `transition` for different HMC samplers.
+### NOTE:  stopping creteria, max_depth::Int, Δ_max::AbstractFloat, n_steps, λ
+###
+
+"""
+Abstract type for termination.
+"""
+abstract type AbstractTermination end
+
+# Termination type for HMC and HMCDA
+struct StaticTermination{D<:AbstractFloat} <: AbstractTermination
+    n_steps :: Int
+    Δ_max :: D
+end
+
+# NoUTurnTermination
+struct NoUTurnTermination{D<:AbstractFloat} <: AbstractTermination
+    max_depth :: Int
+    Δ_max :: D
+    # TODO: add other necessary fields for No-U-Turn stopping creteria.
+end
+
+struct Trajectory{I<:AbstractIntegrator} <: AbstractTrajectory{I}
+    integrator :: I
+    n_steps :: Int # Counter for total leapfrog steps already applied.
+    Δ :: AbstractFloat # Current hamiltonian energy minus starting hamiltonian energy
+    # TODO: replace all ``*Trajectory` types with `Trajectory`.
+    # TODO: add turn statistic, divergent statistic, proposal statistic
+end
+
+isterminated(
```

Comment:
How this function is supposed to use? I mean we can't use it in our HMC implementation as we don't build state one by one but run the integrator with the designated numbers right?

Reply:
Let's leave this as is - I'll fix it in a separate PR.